src/treetensornetwork/abstracttreetensornetwork.jl

```diff
@@ -0,0 +1,399 @@
+  traversal_order = reverse(post_order_dfs_vertices(ψ, root_vertex))
+  return contract(ITensorNetwork(ψ); sequence=traversal_order, kwargs...)
+  # # forward post order edges
+  # ψ = copy(ψ)
+  # for e in post_order_dfs_edges(ψ, root_vertex)
+  #   contract!(ψ, e)
+  # end
+  # return ψ[root_vertex]
+end
+
+function inner(
+  ϕ::AbstractTreeTensorNetwork,
+  ψ::AbstractTreeTensorNetwork;
+  root_vertex=default_root_vertex(ϕ, ψ),
+)
+  ϕᴴ = sim(dag(ψ); sites=[])
+  ψ = sim(ψ; sites=[])
+  ϕψ = ϕᴴ ⊗ ψ
+  # TODO: find the largest tensor and use it as
+  # the `root_vertex`.
+  for e in post_order_dfs_edges(ψ, root_vertex)
+    if has_vertex(ϕψ, 2, src(e)...)
+      ϕψ = contract(ϕψ, (2, src(e)...) => (1, src(e)...))
+    end
+    ϕψ = contract(ϕψ, (1, src(e)...) => (1, dst(e)...))
+    if has_vertex(ϕψ, 2, dst(e)...)
+      ϕψ = contract(ϕψ, (2, dst(e)...) => (1, dst(e)...))
+    end
+  end
+  return ϕψ[1, root_vertex...][]
+end
+
+function norm(ψ::AbstractTreeTensorNetwork)
+  if isortho(ψ)
+    return norm(ψ[only(ortho_center(ψ))])
+  end
+  return √(abs(real(inner(ψ, ψ))))
+end
+
+# 
+# Utility
+# 
+
+function normalize!(ψ::AbstractTreeTensorNetwork)
+  c = ortho_center(ψ)
+  lognorm_ψ = lognorm(ψ)
+  if lognorm_ψ == -Inf
+    return ψ
+  end
+  z = exp(lognorm_ψ / length(c))
+  for v in c
+    ψ[v] ./= z
+  end
+  return ψ
+end
+
+function normalize(ψ::AbstractTreeTensorNetwork)
+  return normalize!(copy(ψ))
+end
+
+function _apply_to_orthocenter!(f, ψ::AbstractTreeTensorNetwork, x)
+  v = first(ortho_center(ψ))
+  ψ[v] = f(ψ[v], x)
+  return ψ
+end
+
+function _apply_to_orthocenter(f, ψ::AbstractTreeTensorNetwork, x)
+  return _apply_to_orthocenter!(f, copy(ψ), x)
+end
+
+Base.:*(ψ::AbstractTreeTensorNetwork, α::Number) = _apply_to_orthocenter(*, ψ, α)
+
+Base.:*(α::Number, ψ::AbstractTreeTensorNetwork) = ψ * α
+
+Base.:/(ψ::AbstractTreeTensorNetwork, α::Number) = _apply_to_orthocenter(/, ψ, α)
+
+Base.:-(ψ::AbstractTreeTensorNetwork) = -1 * ψ
+
+function LinearAlgebra.rmul!(ψ::AbstractTreeTensorNetwork, α::Number)
+  return _apply_to_orthocenter!(*, ψ, α)
+end
+
+function lognorm(ψ::AbstractTreeTensorNetwork)
+  if isortho(ψ)
+    return log(norm(ψ[only(ortho_center(ψ))]))
+  end
+  lognorm2_ψ = logdot(ψ, ψ)
+  rtol = eps(real(scalartype(ψ))) * 10
+  atol = rtol
+  if !IsApprox.isreal(lognorm2_ψ, Approx(; rtol=rtol, atol=atol))
+    @warn "log(norm²) is $lognorm2_T, which is not real up to a relative tolerance of $rtol and an absolute tolerance of $atol. Taking the real part, which may not be accurate."
+  end
+  return 0.5 * real(lognorm2_ψ)
+end
+
+function logdot(ψ1::TTNT, ψ2::TTNT; kwargs...) where {TTNT<:AbstractTreeTensorNetwork}
+  return _log_or_not_dot(ψ1, ψ2, true; kwargs...)
+end
+
+function loginner(ψ1::TTNT, ψ2::TTNT; kwargs...) where {TTNT<:AbstractTreeTensorNetwork}
+  return logdot(ψ1, ψ2; kwargs...)
+end
+
+# TODO: stick with this traversal or find optimal contraction sequence?
+function _log_or_not_dot(
```

Comment:
For code simplicity, it may be better to just implement `logdot`/`loginner` directly, and then just use `inner(x, y) = exp(loginner(x, y))`.

I think I was worried that `exp(loginner(x, y))` might lose some precision but I don't think that's the case in practice.

Reply:
I removed `_log_or_not_dot` in favor of just defining `logdot` but I haven't replaced `inner(...)` with `exp(loginner(...))` everywhere yet. I'll do this later (I still have to define a three-argument logdot for expectation values and sort out some minor things).

Reply:
Sounds good. Also I think it would be better to have the implementation function called `loginner` instead of `logdot` since that is the version people use the most, and it will make it easier to find the implementation.

Reply:
Done.